File: onesky.py
# -*- coding: utf-8 -*-
import hashlib
import requests
import setting
import time
from pprint import pprint
from urlparse import urljoin

class Onesky(object):
    def __init__(self, api_key, api_secret):
        self.api_key = api_key
        self.api_secret = api_secret
        self.api_path = 'https://platform.api.onesky.io/1/'

    def render_auth(self):
        timestamp = str(int(time.time()))
        dev_hash = hashlib.md5()
        dev_hash.update(timestamp)
        dev_hash.update(self.api_secret)
        return {
                'api_key': self.api_key,
                'timestamp': timestamp,
                'dev_hash': dev_hash.hexdigest()
               }

    def api_get(self, path, params=None):
        if params and isinstance(params, dict):
            params.update(self.render_auth())
        else:
            params = self.render_auth()

        return requests.get(urljoin(self.api_path, path), params=params)

    def api_post(self, path, params=None, files=None):
        if params and isinstance(params, dict):
            params.update(self.render_auth())
        else:
            params = self.render_auth()

        return requests.post(urljoin(self.api_path, path), params=params,
                files=files)

    def upload_po(self, project_id, file):
        params = {'file_format': 'GNU_PO'}
        result = self.api_post('projects/%s/files' % project_id, params=params,
                files={'file': file})
        return result.json()

    def download_po(self, project_id, locale, source_file_name, export_file_name=None):
        save_filename = source_file_name
        params = {'locale': locale,
                  'source_file_name': source_file_name}

        if export_file_name:
            params.update(export_file_name=export_file_name)
            save_filename = export_file_name

        result = self.api_get('projects/%s/translations' % project_id, params=params)
        if result.status_code == 200:
            with open('./%s_%s' % (locale, save_filename), 'w') as pof:
                pof.write(result.content)

        return result.status_code

if __name__ == '__main__':
    onesky = Onesky(setting.API_KEY, setting.API_SECRET)
    pprint(onesky.api_get('project-groups').json())
    pprint(onesky.api_get('project-groups/%s/projects' % setting.PROJECT_GROUP_ID).json())
    #with open(setting.PO_FILES, 'rb') as pof:
    #    pprint(onesky.upload_po(setting.PROJECT_ID, pof))
    print onesky.download_po(setting.PROJECT_ID, 'zh-Hant-TW', '...')

File: goonesky/onesky.go
package main

import (
	"bytes"
	"crypto/md5"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"mime/multipart"
	"net/http"
	"net/url"
	"os"
	"path"
	"reflect"
	"time"
)

type AuthData struct {
	ApiKey    string `json:"api_key"`
	Timestamp string `json:"timestamp"`
	Hashkey   string `json:"dev_hash"`
}

func renderAuth(timestamp string) string {
	return fmt.Sprintf("%x", md5.Sum([]byte(fmt.Sprintf("%s%s", timestamp, APISECRET))))
}

func RenderAuth() *AuthData {
	timestamp := fmt.Sprint(time.Now().Unix())
	return &AuthData{
		ApiKey:    APIKEY,
		Timestamp: timestamp,
		Hashkey:   renderAuth(timestamp),
	}
}

func (auth AuthData) ToURLValue() url.Values {
	urlParams := url.Values{}
	p := reflect.ValueOf(auth)
	for i := 0; i < p.NumField(); i++ {
		f := p.Field(i)
		tagName := p.Type().Field(i).Tag.Get("json")
		urlParams.Add(tagName, fmt.Sprint(f))
	}
	return urlParams
}

type OneskyAPI struct{}

var basepath, _ = url.Parse(APIPATH)

func (o OneskyAPI) httpGet(urlPath string) {
	resp, _ := http.Get(urlPath)
	if content, err := ioutil.ReadAll(resp.Body); err == nil {
		fmt.Printf("%s", content)
	}
	defer resp.Body.Close()
}

func (o OneskyAPI) httpPostForm(urlPath string, data url.Values) {
	resp, _ := http.PostForm(urlPath, data)
	if content, err := ioutil.ReadAll(resp.Body); err == nil {
		fmt.Printf("%s", content)
	}
	defer resp.Body.Close()
}

func (o OneskyAPI) httpPostData(urlPath string, data *os.File) {
	//resp, _ := http.Post(urlPath, "multipart/form-data", strings.NewReader(data.Encode()))
	var b bytes.Buffer
	w := multipart.NewWriter(&b)
	fw, err := w.CreateFormFile("file", data.Name())
	if err != nil {
		log.Fatal(err)
	}
	if _, err = io.Copy(fw, data); err != nil {
		return
	}
	w.Close()

	resp, _ := http.Post(urlPath, w.FormDataContentType(), bytes.NewReader(b.Bytes()))
	if content, err := ioutil.ReadAll(resp.Body); err == nil {
		fmt.Printf("%s", content)
	}
	defer resp.Body.Close()
}

func (o OneskyAPI) UploadPO(params *AuthData, files ...string) {
	urlParams := params.ToURLValue()
	urlParams.Add("file_format", "GNU_PO")
	basepath.Path = path.Join(APIVERSION, "projects", PROJECTID, "files")
	basepath.RawQuery = urlParams.Encode()

	for _, filename := range files {
		file, err := os.Open(filename)
		if err != nil {
			log.Fatalf("`%s` not find.", filename)
		}
		o.httpPostData(basepath.String(), file)
	}
}

func (o OneskyAPI) GetProjectInfo(params *AuthData) {
	basepath.Path = path.Join(APIVERSION, "projects", PROJECTID, "languages")
	basepath.RawQuery = params.ToURLValue().Encode()
	o.httpGet(basepath.String())
}

func (o OneskyAPI) GetFilesList(params *AuthData) {
	basepath.Path = path.Join(APIVERSION, "projects", PROJECTID, "files")
	basepath.RawQuery = params.ToURLValue().Encode()
	o.httpGet(basepath.String())
}

func main() {
	auth := RenderAuth()
	fmt.Println(auth)
	o := OneskyAPI{}
	o.GetProjectInfo(auth)
	o.GetFilesList(auth)

	o.UploadPO(auth, "onesky.po", "test.po")
}
